src/MasonryFlashList.tsx

```diff
@@ -0,0 +1,409 @@
+              }
+              drawDistance={drawDistance}
+              estimatedListSize={{
+                height: estimatedListSize.height,
+                width:
+                  ((getListRenderedSize(parentFlashList)?.width ||
+                    estimatedListSize.width) /
+                    columnCount) *
+                  (getColumnSizeMultiplier?.(
+                    args.item,
+                    args.index,
+                    columnCount,
+                    props.extraData
+                  ) ?? 1),
+              }}
+              extraData={props.extraData}
+              CellRendererComponent={CellRendererComponent}
+              ItemSeparatorComponent={ItemSeparatorComponent}
+              viewabilityConfig={viewabilityConfig}
+              onViewableItemsChanged={
+                onViewableItemsChanged
+                  ? (info) => {
+                      updateViewToken(info.viewableItems);
+                      updateViewToken(info.changed);
+                      onViewableItemsChanged?.(info);
+                    }
+                  : undefined
+              }
+              overrideItemLayout={
+                overrideItemLayout
+                  ? (layout, item, index, _, extraData) => {
+                      overrideItemLayout?.(
+                        layout,
+                        item.originalItem,
+                        item.originalIndex,
+                        columnCount,
+                        extraData
+                      );
+                      layout.span = undefined;
+                    }
+                  : undefined
+              }
+            />
+          );
+        }}
+        overrideItemLayout={
+          getColumnSizeMultiplier
+            ? (layout, item, index, maxColumns, extraData) => {
+                layout.span = getColumnSizeMultiplier?.(
+                  item,
+                  index,
+                  maxColumns,
+                  extraData
+                );
+              }
+            : undefined
+        }
+      />
+    );
+  }
+);
+
+/**
+ * Splits data for each column's FlashList
+ */
+const useDataSet = <T,>(
+  columnCount: number,
+  autoOptimizeLayout: boolean,
+  sourceData?: FlashListProps<T>["data"],
+  overrideItemLayout?: MasonryFlashListProps<T>["overrideItemLayout"],
+  extraData?: MasonryFlashListProps<T>["extraData"]
+): MasonryListItem<T>[][] => {
+  return useMemo(() => {
+    if (!sourceData || sourceData.length === 0) {
+      return [];
+    }
+    const columnHeightTracker = new Array<number>(columnCount).fill(0);
+    const layoutObject: { size: number | undefined } = { size: undefined };
+    const dataSet = new Array<MasonryListItem<T>[]>(columnCount);
+    const dataSize = sourceData.length;
+
+    for (let i = 0; i < columnCount; i++) {
+      dataSet[i] = [];
+    }
+    for (let i = 0; i < dataSize; i++) {
+      let nextColumnIndex = i % columnCount;
+      if (autoOptimizeLayout) {
```

Comment:
While I agree in general my main concern is with implicitly changing the order of items. I referred to Android's StaggeredGridLayoutManager and it keeps optimize on by default. The difference is that it makes sure that you, visually, see the optimization happen and we can't do that here. Animations across columns aren't possible in the current design. Making it apparent is a theme, even SwiftUI libraries do that same.

My worry is that by doing it implicitly we will confuse people as other open source packages in RN don't do this at all. Folks debugging the data order is one possible side effect.

Also, overrideItemLayout isn't trivial to implement. It increases the entry barrier by a lot (almost as much as RLV). It might be better to let people turn it on if they need it. I also wonder if this is a real concern when pagination is involved.

I can see the point but reorder not being apparent is a problem. Let me think about this more otherwise we can just wait to see how people use it and make the change.

I had also considered a new prop for this. Something simpler to implement but introducing one more thing to learn didn't seem very ideal.